src/metadataExt.ts

```diff
@@ -0,0 +1,261 @@
+// (C) 2020 GoodData Corporation
+import { MetadataModule } from "./metadata";
+import { UserModule } from "./user";
+import { XhrModule } from "./xhr";
+import { IGetObjectUsingOptions } from "./interfaces";
+import cloneDeepWith from "lodash/cloneDeepWith";
+
+export interface ICopyDashboardOptions {
+    keepReferences: boolean;
+    name?: string;
+}
+
+export class MetadataModuleExt {
+    private metadataModule: MetadataModule;
+    private userModule: UserModule;
+    private xhr: XhrModule;
+
+    constructor(xhr: XhrModule) {
+        this.xhr = xhr;
+        this.metadataModule = new MetadataModule(xhr);
+        this.userModule = new UserModule(xhr);
+    }
+
+    /**
+     * Creates a clone of a dashoard with given uri
+     *
+     * @method saveAs
+     * @param {string} projectID id of the project
+     * @param {string} dashboardUri uri of the dashboard
+     * @param {ICopyDashboardOptions} options object with options:
+     *      - keepReferences {boolean} choose whether dashboard is cloned with same references
+     *      - name {string} new name for dashboard
+     * @return {string} uri of cloned dashboard
+     */
+    public async saveAs(
+        projectID: string,
+        dashboardUri: string,
+        options: ICopyDashboardOptions,
+    ): Promise<string> {
+        const dashboard = await this.metadataModule.getObjectDetails(dashboardUri);
+        const { analyticalDashboard } = await this.metadataModule.createObject(projectID, {
+            ...dashboard,
+            analyticalDashboard: {
+                ...dashboard.analyticalDashboard,
+                meta: {
+                    ...dashboard.analyticalDashboard.meta,
+                    title: this.getDashboardName(dashboard.analyticalDashboard.meta.title, options.name),
+                },
+            },
+        });
+
+        const dashboardID: string = analyticalDashboard.meta.uri.split("/")[5];
```

Comment:
I think there might already by a function in gd-js to obtain object ID from URI. please look around, it could be somewhere around utils; if the function does not exist then I suggest to create one and use it in all places where we need to get object ID from URL.

Reply:
ideally do something like getObjectIdForObj(obj).. and the getObjectId() looks whether obj.meta.uri is available and if so do whatever